File: sdk/typescript/src/namespaces/feedback.ts
/**
 * Feedback Namespace API
 *
 * Provides endpoints for collecting user feedback including:
 * - NPS (Net Promoter Score) surveys
 * - Feature feedback
 * - Bug reports
 * - General suggestions
 */

/**
 * Types of feedback.
 */
export type FeedbackType =
  | 'nps'
  | 'feature_request'
  | 'bug_report'
  | 'general'
  | 'debate_quality';

/**
 * NPS submission request.
 */
export interface NPSSubmission {
  /** Score from 0-10 */
  score: number;
  /** Optional comment explaining the score */
  comment?: string;
  /** Optional metadata context */
  context?: Record<string, unknown>;
}

/**
 * General feedback submission.
 */
export interface FeedbackSubmission {
  /** Type of feedback */
  type: FeedbackType;
  /** Required comment/description */
  comment: string;
  /** Optional rating score */
  score?: number;
  /** Optional metadata context */
  context?: Record<string, unknown>;
}

/**
 * Feedback submission response.
 */
export interface FeedbackResponse {
  success: boolean;
  feedback_id: string;
  message: string;
}

/**
 * NPS summary statistics.
 */
export interface NPSSummary {
  /** Overall NPS score (-100 to 100) */
  nps_score: number;
  /** Total number of responses */
  total_responses: number;
  /** Count of promoters (9-10 scores) */
  promoters: number;
  /** Count of passives (7-8 scores) */
  passives: number;
  /** Count of detractors (0-6 scores) */
  detractors: number;
  /** Time period in days */
  period_days: number;
}

/**
 * Feedback prompt configuration.
 */
export interface FeedbackPrompt {
  type: string;
  question: string;
  scale?: {
    min: number;
    max: number;
    labels: Record<string, string>;
  };
  follow_up?: string;
}

/**
 * Feedback API for collecting user feedback.
 */
export class FeedbackAPI {
  private baseUrl: string;
  private headers: HeadersInit;

  constructor(baseUrl: string, apiKey: string) {
    this.baseUrl = baseUrl;
    this.headers = {
      'Content-Type': 'application/json',
      Authorization: `Bearer ${apiKey}`,
    };
  }

  /**
   * Submit NPS feedback.
   *
   * @param submission - NPS score and optional comment
   */
  async submitNPS(submission: NPSSubmission): Promise<FeedbackResponse> {
    if (submission.score < 0 || submission.score > 10) {
      throw new Error('NPS score must be between 0 and 10');
    }

    const response = await fetch(`${this.baseUrl}/api/v1/feedback/nps`, {
      method: 'POST',
      headers: this.headers,
      body: JSON.stringify(submission),
    });
    if (!response.ok) throw new Error(`Failed to submit NPS: ${response.statusText}`);
    return response.json();
  }

  /**
   * Submit general feedback.
   *
   * @param submission - Feedback type and comment
   */
  async submitFeedback(submission: FeedbackSubmission): Promise<FeedbackResponse> {
    if (!submission.comment) {
      throw new Error('Comment is required for feedback submission');
    }

    const response = await fetch(`${this.baseUrl}/api/v1/feedback/general`, {
      method: 'POST',
      headers: this.headers,
      body: JSON.stringify(submission),
    });
    if (!response.ok) throw new Error(`Failed to submit feedback: ${response.statusText}`);
    return response.json();
  }

  /**
   * Submit a feature request.
   *
   * @param description - Description of the feature
   * @param context - Optional context metadata
   */
  async submitFeatureRequest(
    description: string,
    context?: Record<string, unknown>
  ): Promise<FeedbackResponse> {
    return this.submitFeedback({
      type: 'feature_request',
      comment: description,
      context,
    });
  }

  /**
   * Submit a bug report.
   *
   * @param description - Description of the bug
   * @param context - Optional context (steps to reproduce, etc.)
   */
  async submitBugReport(
    description: string,
    context?: Record<string, unknown>
  ): Promise<FeedbackResponse> {
    return this.submitFeedback({
      type: 'bug_report',
      comment: description,
      context,
    });
  }

  /**
   * Submit debate quality feedback.
   *
   * @param debateId - ID of the debate
   * @param comment - Quality feedback comment
   * @param score - Optional quality score
   */
  async submitDebateQualityFeedback(
    debateId: string,
    comment: string,
    score?: number
  ): Promise<FeedbackResponse> {
    return this.submitFeedback({
      type: 'debate_quality',
      comment,
      score,
      context: { debate_id: debateId },
    });
  }

  /**
   * Get NPS summary statistics (admin only).
   *
   * @param days - Number of days to include (default 30)
   */
  async getNPSSummary(days: number = 30): Promise<NPSSummary> {
    const response = await fetch(
      `${this.baseUrl}/api/v1/feedback/nps/summary?days=${days}`,
      {
        method: 'GET',
        headers: this.headers,
      }
    );
    if (!response.ok) throw new Error(`Failed to get NPS summary: ${response.statusText}`);
    return response.json();
  }

  /**
   * Get active feedback prompts for the current user.
   */
  async getPrompts(): Promise<{ prompts: FeedbackPrompt[] }> {
    const response = await fetch(`${this.baseUrl}/api/v1/feedback/prompts`, {
      method: 'GET',
      headers: this.headers,
    });
    if (!response.ok) throw new Error(`Failed to get prompts: ${response.statusText}`);
    return response.json();
  }
}

File: sdk/python/aragora/namespaces/feedback.py
"""
Feedback Namespace API

Provides methods for collecting user feedback including NPS surveys,
feature requests, bug reports, and general suggestions.

Features:
- NPS (Net Promoter Score) submission
- General feedback submission
- NPS analytics (admin)
- Feedback prompts
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any, Literal

if TYPE_CHECKING:
    from ..client import AragoraAsyncClient, AragoraClient


FeedbackType = Literal["feature_request", "bug_report", "general", "debate_quality"]


class FeedbackAPI:
    """
    Synchronous Feedback API.

    Provides methods for submitting and managing user feedback:
    - NPS surveys
    - Feature requests
    - Bug reports
    - General feedback
    - Feedback analytics (admin)

    Example:
        >>> client = AragoraClient(base_url="https://api.aragora.ai", api_key="...")
        >>> client.feedback.submit_nps(score=9, comment="Great product!")
        >>> prompts = client.feedback.get_prompts()
    """

    def __init__(self, client: AragoraClient):
        self._client = client

    # ===========================================================================
    # NPS (Net Promoter Score)
    # ===========================================================================

    def submit_nps(
        self,
        score: int,
        comment: str | None = None,
        context: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        """
        Submit NPS (Net Promoter Score) feedback.

        Args:
            score: NPS score from 0-10
                   - 0-6: Detractor
                   - 7-8: Passive
                   - 9-10: Promoter
            comment: Optional comment explaining the score
            context: Optional metadata (e.g., feature being used)

        Returns:
            Dict with success status and feedback_id

        Raises:
            ValidationError: If score is not 0-10
        """
        data: dict[str, Any] = {"score": score}
        if comment:
            data["comment"] = comment
        if context:
            data["context"] = context

        return self._client.request("POST", "/api/v1/feedback/nps", json=data)

    def get_nps_summary(self, days: int = 30) -> dict[str, Any]:
        """
        Get NPS summary analytics (admin only).

        Args:
            days: Number of days to include in summary (default: 30)

        Returns:
            Dict with:
            - nps_score: Overall NPS (-100 to 100)
            - total_responses: Number of responses
            - promoters: Count of 9-10 scores
            - passives: Count of 7-8 scores
            - detractors: Count of 0-6 scores
            - period_days: Period covered

        Requires: feedback.update permission (admin)
        """
        return self._client.request(
            "GET",
            "/api/v1/feedback/nps/summary",
            params={"days": days},
        )

    # ===========================================================================
    # General Feedback
    # ===========================================================================

    def submit_feedback(
        self,
        comment: str,
        feedback_type: FeedbackType = "general",
        score: int | None = None,
        context: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        """
        Submit general feedback.

        Args:
            comment: Feedback content (required)
            feedback_type: Type of feedback:
                - "feature_request": New feature idea
                - "bug_report": Bug or issue
                - "general": General feedback
                - "debate_quality": Feedback on debate quality
            score: Optional rating score
            context: Optional metadata

        Returns:
            Dict with success status and feedback_id
        """
        data: dict[str, Any] = {
            "type": feedback_type,
            "comment": comment,
        }
        if score is not None:
            data["score"] = score
        if context:
            data["context"] = context

        return self._client.request("POST", "/api/v1/feedback/general", json=data)

    def submit_feature_request(
        self,
        comment: str,
        context: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        """
        Submit a feature request.

        Convenience method for submit_feedback with type="feature_request".

        Args:
            comment: Description of the requested feature
            context: Optional metadata

        Returns:
            Dict with success status and feedback_id
        """
        return self.submit_feedback(
            comment=comment,
            feedback_type="feature_request",
            context=context,
        )

    def submit_bug_report(
        self,
        comment: str,
        context: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        """
        Submit a bug report.

        Convenience method for submit_feedback with type="bug_report".

        Args:
            comment: Description of the bug
            context: Optional metadata (e.g., steps to reproduce)

        Returns:
            Dict with success status and feedback_id
        """
        return self.submit_feedback(
            comment=comment,
            feedback_type="bug_report",
            context=context,
        )

    # ===========================================================================
    # Feedback Prompts
    # ===========================================================================

    def get_prompts(self) -> dict[str, Any]:
        """
        Get active feedback prompts for the user.

        Returns prompts based on user activity and timing.

        Returns:
            Dict with prompts array containing question configs
        """
        return self._client.request("GET", "/api/v1/feedback/prompts")


class AsyncFeedbackAPI:
    """
    Asynchronous Feedback API.

    Example:
        >>> async with AragoraAsyncClient(base_url="https://api.aragora.ai") as client:
        ...     await client.feedback.submit_nps(score=9, comment="Great!")
        ...     prompts = await client.feedback.get_prompts()
    """

    def __init__(self, client: AragoraAsyncClient):
        self._client = client

    # ===========================================================================
    # NPS (Net Promoter Score)
    # ===========================================================================

    async def submit_nps(
        self,
        score: int,
        comment: str | None = None,
        context: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        """Submit NPS (Net Promoter Score) feedback."""
        data: dict[str, Any] = {"score": score}
        if comment:
            data["comment"] = comment
        if context:
            data["context"] = context

        return await self._client.request("POST", "/api/v1/feedback/nps", json=data)

    async def get_nps_summary(self, days: int = 30) -> dict[str, Any]:
        """Get NPS summary analytics (admin only)."""
        return await self._client.request(
            "GET",
            "/api/v1/feedback/nps/summary",
            params={"days": days},
        )

    # ===========================================================================
    # General Feedback
    # ===========================================================================

    async def submit_feedback(
        self,
        comment: str,
        feedback_type: FeedbackType = "general",
        score: int | None = None,
        context: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        """Submit general feedback."""
        data: dict[str, Any] = {
            "type": feedback_type,
            "comment": comment,
        }
        if score is not None:
            data["score"] = score
        if context:
            data["context"] = context

        return await self._client.request("POST", "/api/v1/feedback/general", json=data)

    async def submit_feature_request(
        self,
        comment: str,
        context: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        """Submit a feature request."""
        return await self.submit_feedback(
            comment=comment,
            feedback_type="feature_request",
            context=context,
        )

    async def submit_bug_report(
        self,
        comment: str,
        context: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        """Submit a bug report."""
        return await self.submit_feedback(
            comment=comment,
            feedback_type="bug_report",
            context=context,
        )

    # ===========================================================================
    # Feedback Prompts
    # ===========================================================================

    async def get_prompts(self) -> dict[str, Any]:
        """Get active feedback prompts for the user."""
        return await self._client.request("GET", "/api/v1/feedback/prompts")
